### PR TITLE
[App] Multiprocessing-safe work pickling

### PR DESCRIPTION
**Pull request by hhsecond**
## What does this PR do?

For LightningWork to become more abstract components, we'd need to enable the work to be pickled/unpickled within the framework. This PR introduces the multiprocessing-safe, cloud-safe work pickling mechanism.

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/utilities/safe_pickle.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    for w in _LightningAppRef.get_current().works:
+        if work.name == w.name:
+            # copying the work object to avoid modifying the original work object
+            copied_work = deepcopy(w)
+            break
+    else:
+        raise ValueError(f"Work with name {work.name} not found in the app references")
+    # if work is defined in the __main__ or __mp__main__ (the entrypoint file for `lightning run app` command),
+    # pickling/unpickling will fail, hence we need patch the module information
+    if "_main__" in copied_work.__class__.__module__:
```
Comment (Borda) [inline src/lightning_app/utilities/safe_pickle.py:41]:
```suggestion
    if "__main__" in copied_work.__class__.__module__:
```

File: src/lightning_app/utilities/safe_pickle.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        relative_path = Path(work_class_module.__file__).relative_to(Path.cwd())
+        expected_module_name = relative_path.as_posix().replace(".py", "").replace("/", ".")
+        # TODO @sherin: also check if the module is importable from the CWD
+        fake_module = types.ModuleType(expected_module_name)
+        fake_module.__dict__.update(work_class_module.__dict__)
+        fake_module.__dict__["__name__"] = expected_module_name
+        sys.modules[expected_module_name] = fake_module
+        for k, v in fake_module.__dict__.items():
+            if not k.startswith("__") and hasattr(v, "__module__"):
+                if "_main__" in v.__module__:
```
Comment (Borda) [inline src/lightning_app/utilities/safe_pickle.py:52]:
```suggestion
                if "__main__" in v.__module__:
```

File: tests/tests_app/utilities/test_safe_pickle.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import subprocess
+from pathlib import Path
+
+
+def test_safe_pickle_app():
+    test_dir = Path(__file__).parent / "testdata"
+    proc = subprocess.Popen(
+        ["lightning", "run", "app", "safe_pickle_app.py", "--open-ui", "false"], stdout=subprocess.PIPE, cwd=test_dir
+    )
+    assert "Exiting the pickling app successfully!!" in proc.stdout.read().decode("UTF-8")
```
Comment (Borda) [inline tests/tests_app/utilities/test_safe_pickle.py:10]:
```suggestion
    assert "Exiting the pickling app successfully!" in proc.stdout.read().decode("UTF-8")
```

File: tests/tests_app/utilities/testdata/safe_pickle_app.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        # deleting the script
+        Path("script_that_loads_pickled_work.py").unlink()
+        # deleting the pkl file
+        Path("work.pkl").unlink()
+
+        self._exit("Exiting the pickling app successfully!!")
+
+
+app = LightningApp(RootFlow())
```
Comment (Borda) [inline tests/tests_app/utilities/testdata/safe_pickle_app.py:63]: wrap this to a test

**Comment by Borda** [comment]
@hhsecond could you pls fill checklist? :)

File: src/lightning_app/utilities/safe_pickle.py (hunk 1 above)
Comment (hhsecond) [reply src/lightning_app/utilities/safe_pickle.py:45]: that doesn't work `multiprocessing.Process` creates the main module as `__mp_main__`

File: src/lightning_app/utilities/safe_pickle.py (hunk 2 above)
Comment (hhsecond) [reply src/lightning_app/utilities/safe_pickle.py:62]: that doesn't work multiprocessing.Process creates the main module as __mp_main__

File: tests/tests_app/utilities/testdata/safe_pickle_app.py (hunk 4 above)
Comment (hhsecond) [reply tests/tests_app/utilities/testdata/safe_pickle_app.py:63]: what do you mean?  don't keep it in `testdata`?

File: src/lightning_app/utilities/safe_pickle.py (hunk 1 above)
Comment (Borda) [reply src/lightning_app/utilities/safe_pickle.py:45]: ahh, I see, can we add it then as a comment for later me? :)

File: tests/tests_app/utilities/testdata/safe_pickle_app.py (hunk 4 above)
Comment (Borda) [reply tests/tests_app/utilities/testdata/safe_pickle_app.py:63]:
```suggestion
def test_app_with_root_flow():
    app = LightningApp(RootFlow())
```

File: src/lightning_app/CHANGELOG.md [hunk 5]
```diff
@@ -14,6 +14,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Show a message when `BuildConfig(dockerfile="...")` is passed but a `Dockerfile` file is already present in the Work ([#15799](https://github.com/Lightning-AI/lightning/pull/15799))
 
 - Added a CloudMultiProcessBackend which enables running a child App from within the Flow in the cloud ([#15800](https://github.com/Lightning-AI/lightning/pull/15800))
+- Utility for pickling work object safely even from a child process ([#15836](https://github.com/Lightning-AI/lightning/pull/15836))
```
Comment (ethanwharris) [inline src/lightning_app/CHANGELOG.md:17]:
```suggestion

- Utility for pickling work object safely even from a child process ([#15836](https://github.com/Lightning-AI/lightning/pull/15836))
```

File: src/lightning_app/utilities/safe_pickle.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import typing
+from copy import deepcopy
+from pathlib import Path
+
+from lightning_app.core.work import LightningWork
+from lightning_app.utilities.app_helpers import _LightningAppRef
+
+
+def get_picklable_work(work: LightningWork) -> LightningWork:
+    """Pickling a LightningWork instance fails if it doesn from the work process
```
Comment (ethanwharris) [inline src/lightning_app/utilities/safe_pickle.py:13]:
```suggestion
    """Pickling a LightningWork instance fails if done from the work process
```

File: src/lightning_app/utilities/safe_pickle.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            if not k.startswith("__") and hasattr(v, "__module__"):
+                if "_main__" in v.__module__:
+                    v.__module__ = expected_module_name
+
+    # removing reference to backend; backend is not picklable because of openapi client reference in it
+    copied_work._backend = None
+    return copied_work
+
+
+def dump(work: LightningWork, f: typing.BinaryIO) -> None:
```
Comment (ethanwharris) [inline src/lightning_app/utilities/safe_pickle.py:70]: Maybe these utilitiies should be added to the API reference or something, currently not really discoverable

**Review by ethanwharris — APPROVED** [review]
LGTM, small comments

File: src/lightning_app/utilities/safe_pickle.py (hunk 1 above)
Comment (hhsecond) [reply src/lightning_app/utilities/safe_pickle.py:45]: I think the comment is already there. Just above this line :)

File: tests/tests_app/utilities/testdata/safe_pickle_app.py (hunk 4 above)
Comment (hhsecond) [reply tests/tests_app/utilities/testdata/safe_pickle_app.py:63]: basically, my test is to run this file. Do you see any problems with keeping the `app` in the global scope?

Comment (Borda) [reply tests/tests_app/utilities/testdata/safe_pickle_app.py:63]: ok, can you point me out where is this file called? :)

Comment (hhsecond) [reply tests/tests_app/utilities/testdata/safe_pickle_app.py:63]: https://github.com/Lightning-AI/lightning/blob/safe_pickle/tests/tests_app/utilities/test_safe_pickle.py

File: src/lightning_app/utilities/safe_pickle.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        work_class_module = sys.modules[copied_work.__class__.__module__]
+        work_class_file = work_class_module.__file__
+        if not work_class_file:
+            raise ValueError(
+                f"Cannot pickle work class {copied_work.__class__.__name__} because we "
+                f"couldn't identify the module file"
+            )
+        relative_path = Path(work_class_module.__file__).relative_to(Path.cwd())  # type: ignore
+        expected_module_name = relative_path.as_posix().replace(".py", "").replace("/", ".")
+        # TODO @sherin: also check if the module is importable from the CWD
```
Comment (hhsecond) [inline src/lightning_app/utilities/safe_pickle.py:71]:
```suggestion
```

File: src/lightning_app/utilities/safe_pickle.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        └── foo
+            ├── __init__.py
+            └── bar
+                └── app.py
+    """
+
+    # pickling the user work class - pickling `self` will cause issue because the
+    # work is running under a process, in local
+    app_ref = _LightningAppRef.get_current()
+    if app_ref is None:
```
[the remark this answers is not in the document]
Comment (hhsecond) [reply src/lightning_app/utilities/safe_pickle.py:48]: added a comment

**Pull request merged**